### ISSUE TITLE
un.rpyc (1.0.8) fails to load with Ren'py 7.0 games

### ISSUE THREAD

**Issue by OrdosAlpha** [CensoredUsername/unrpyc#74]
Tried decompiling the .rpyc files of several games compiled using Ren'Py 7.0, and un.rpyc fails to load. Works perfectly with games compiled with Ren'Py 6.99.14.3 and older.

Included the tracebacks for two of the games I've tried:- 
"Full traceback:
  File "F:\Renpy\Summertime Saga 0.16\renpy\bootstrap.py", line 306, in bootstrap
    renpy.main.main()
  File "F:\Renpy\Summertime Saga 0.16\renpy\main.py", line 375, in main
    renpy.game.script.load_script()  # sets renpy.game.script.
  File "F:\Renpy\Summertime Saga 0.16\renpy\script.py", line 266, in load_script
    self.load_appropriate_file(".rpyc", ".rpy", dir, fn, initcode)
  File "F:\Renpy\Summertime Saga 0.16\renpy\script.py", line 749, in load_appropriate_file
    raise Exception("Could not load file %s." % lastfn)
Exception: Could not load file F:\Renpy\Summertime Saga 0.16/game/un.rpyc.

Windows-7-6.1.7601-SP1
Ren'Py 7.0.0.196
SummertimeSaga 0.16.0
Fri Jul 27 01:16:02 2018"

"Full traceback:
  File "F:\Renpy\A Town Uncovered 0.19b\renpy\bootstrap.py", line 306, in bootstrap
    renpy.main.main()
  File "F:\Renpy\A Town Uncovered 0.19b\renpy\main.py", line 357, in main
    renpy.game.script.load_script()  # sets renpy.game.script.
  File "F:\Renpy\A Town Uncovered 0.19b\renpy\script.py", line 266, in load_script
    self.load_appropriate_file(".rpyc", ".rpy", dir, fn, initcode)
  File "F:\Renpy\A Town Uncovered 0.19b\renpy\script.py", line 749, in load_appropriate_file
    raise Exception("Could not load file %s." % lastfn)
Exception: Could not load file F:\Renpy\A Town Uncovered 0.19b/game/un.rpyc.

Windows-7-6.1.7601-SP1
Ren'Py 7.0.0.196
 
Fri Jul 27 01:17:59 2018"

**Comment by CensoredUsername** [comment]
That's odd. I can't find a recent change in ren'py that woud cause this to happen. I'll try to investigate.

**Comment by PMiley** [comment]
I had this issue with Summertime Saga 0.16.0 and 0.16.1 but not A Town Uncovered 0.19b. I did notice that Summertime Saga adds several additional Python libraries. 
![summertime libraries](https://user-images.githubusercontent.com/16994318/43558259-5fd08aa8-95bd-11e8-9da8-15beb8754e09.png)

**Comment by annhwi** [comment]
pretty much the same story here w/ Summertime Saga 16.1.  Unpacking the .rpa files w/ unren is no issue... but traceback is virtually identical for the un.rpyc.

[traceback.txt](https://github.com/CensoredUsername/unrpyc/files/2255688/traceback.txt)

**Comment by johngary1951** [comment]
Get the new version of UnRen.bat here https://f95zone.com/threads/unren-bat-v0-7-rpa-extractor-rpyc-decompiler-console-developer-menu-enabler.3083/

**Comment by davidosky101** [comment]
this is a modification of un.rpyc (1.0.8) where they fixed the error you mention, unfortunately they couldn't fix it to work with renpy 7 and the new layered image system, hopefully the author can fix it.
[UnRen-v0.7.zip](https://github.com/CensoredUsername/unrpyc/files/2635441/UnRen-v0.7.zip)

**Comment by CensoredUsername** [comment]
First of all, UnRen does not use the un.rpyc file.

Second, it doesn't fix the mentioned error, it just uses a modification of normal unrpyc that purely removes multiprocessing support so it can be ran using ren'py's python executable instead of using python installed on the system.

Third, UnRen violates the license of this project (it does not include the license disclaimer anywhere as is required) so I'm not too keen on people mentioning it here. Especially since it is little more than a repackage of unrpyc and rpatool.

Since it doesn't fix the error I'll mark advertisements of it as spam until they fix the license issue.

**Comment by CensoredUsername** [comment]
@OrdosAlpha I've found one possible source of the bug and fixed it. Could you check if your problems are resolved?

**Comment by skr68** [comment]
Hi,

first of all, thanks for this useful tool. However, since I also experience this problem, I would like to pick up here. I downloaded un.rpyc v1.0.9 and tried to decompile a file containing a "layered image" from game Summertime Saga v0.17.5. It failed as before, with just the line "layered image <name>:"

No errors were given in the log. Is there something I can do to help narrowing this?

**Comment by CensoredUsername** [comment]
@skr68 that's a different bug, that's just that layered images aren't supported yet in decompilation.

**Comment by CensoredUsername** [comment]
@srk68 Should be supported now.

And as I've gotten no further info I'm closing this issue as I cannot reproduce it myself.

**Comment by srk68** [comment]
Hi,

I don't know if a closed issue can be re-opened. But I just tested Summertime Saga with the new 1.0.11 version of un.rpyc, and the error as mentioned in the initial post is still there.

I remember there was a workaround for such problems in the past: deleting the locations and minigames folder in the game\scripts folder before running un.rpyc. I did that and it worked, of course without the deleted folders. I then restored those folders, started the game (with un.rpyc) again, and finally it managed to decompile everything.

Maybe this information helps?

**Comment by srk68** [comment]
Hi!

I just checked and it is the same traceback as in the first post (apart from path names and dates).

I tried narrowing the problem down, but it appears that it is not a specific file or folder that needs to be removed:

I first deleted half the folders under game\scripts\locations, and it failed to decompile. I restored the folders and deleted the other half, again it failed.

Then I deleted one folder after another until it succeeded. But the last folder I deleted was not the problem because when I restored it and deleted all the other folders, everything went well (which was to be expected because everything goes well, too, when I restore all deleted folders after the first decompilation went through for the rest, as I described before).

So, basically, it can decompile everything, but not everything together in one pass.

**Comment by CensoredUsername** [comment]
@skr68 [here you go](https://www.dropbox.com/s/7mhlzsnwi1zj22c/un.rpyc). It does nothing special except for printing a bunch of stuff to the game's log.txt. If you could forward any output in log.txt, unrpyc.log.txt, or other log files with information on the crash to me I'd be very thankful.

**Comment by srk68** [comment]
Hi,

I just tried with version 0.17.1 and received the same error. Please find the log file attached.

[log.zip](https://github.com/CensoredUsername/unrpyc/files/2726769/log.zip)

**Comment by srk68** [comment]
Hi again,

I think I figured out what the problem was....

I first created decompiled the RPA files with rpatool to get all the RPYC and PNG files, then (to save space) I deleted the RPA files. Only after this I put un.rpyc into the game directory to decompile the RPYC files. This fails as described above.

I was not aware that it is not necessary to create the RPYC files before. If I just start the game with un.rpyc without prior rpatool, it manages to create the RPY files from the RPA files directly.

**Comment by CensoredUsername** [comment]
I've found the issue: apparently those games simply have more script files than file handles the OS support. I wasn't noticing it as this issue is only encountered if you use un.rpyc after extracting the script files from the archives manually. It should be fixed now.

**Comment by srk68** [comment]
OK, tested with un.rpyc v1.0.12, works perfectly now!

Thanks!
